### PR TITLE
Use pipe to communicate between client and server at startup

### LINKED ISSUE #1142: Mosh clears the terminal and adds an empty lines on login
server: Debian 10, mosh 1.3.2 [build mosh 1.3.2]
"ssh server" gives me this output:
```
[user@desktop ~]$ ssh server
user@server:~$
```
"mosh server" completely clears the screen and adds three additional empty lines:
```
  
  
user@server:~$
```

I don't mind clearing the screen, but I would like to get rid of the empty lines.

sshd_config:
  Port *redacted*
  PermitRootLogin no
  PubkeyAuthentication yes
  PasswordAuthentication no
  PermitEmptyPasswords no
  ChallengeResponseAuthentication no
  UsePAM no
  PrintMotd no
  PrintLastLog no

file /etc/motd
/etc/motd: empty

server:
echo $TERM
rxvt-unicode-256color

client:
echo $TERM
rxvt-unicode-256color

terminal used: rxvt-unicode with 256 color patch applied

other terminal tested - xterm, same behaviour

### PR DESCRIPTION
**Pull request by ahedberg**
Fixes #1142

### CODE REVIEW

File: src/frontend/mosh-server.cc [hunk 1]
```diff
@@ -592,8 +603,12 @@ static int run_server( const char *desired_ip, const char *desired_port,
 
     /* Wait for parent to release us. */
     char linebuf[81];
-    if (fgets(linebuf, sizeof linebuf, stdin) == NULL) {
-      err( 1, "parent signal" );
+    // -1, errno == EINTR -- retry
+    // otherwise, give up
+    while ( read( pipes[0], linebuf, sizeof( linebuf ) ) == -1 ) {
```
Comment (achernya) [inline src/frontend/mosh-server.cc:608]: Am I understanding correctly that before, we tried to call fgets() until we read a newline, then did execve, but now, we're going to do read() until the other end of the pipe is closed (i.e., EOF)? 

That means, AIUI, if we do get an error here and it's not EINTR, it's fatal?

**Review by achernya — COMMENTED** [review]
Can we augment the git commit description with how this commit works? The PR description won't show up in the git log

File: src/frontend/mosh-server.cc (hunk 1 above)
Comment (ahedberg) [reply src/frontend/mosh-server.cc:608]: Prior to this change, I believe it was possible for the `fgets()` until we read a newline to block forever, in the case where the newline was never received. With this change, that hang at startup becomes a fatal error instead. The fatal error seems preferable to me, but I might be misunderstanding what the original code was doing.

**Comment by cgull** [comment]
> Fixes #1142

I don't think this fixes anything?  In the prior mosh-server code, that read/write on the pty are before any user data passes across it, the newline should never be seen by the user.  The problems with excess newlines are in mosh.pl, at least in my experience.

I am concerned about the additional complexity added by this commit.  At the very least, `CLOEXEC` should be set on the pipe fds.

**Comment by achernya** [comment]
>  that read/write on the pty are before any user data passes across it, the newline should never be seen by the user.

Empirically, it *is* being seen by the user.

**Comment by achernya** [comment]
@ahedberg I fixed the CI config, I believe if you rebase and push CI should be clean.

File: src/frontend/mosh-server.cc (hunk 1 above)
Comment (achernya) [reply src/frontend/mosh-server.cc:609]: Discussed offline, this is not a behavior change, there was a fatal error on fgets hitting EOF before and there still is now.

**Review by achernya — APPROVED** [review]

**Pull request merged**